client/heartbeatstop.go

```diff
@@ -0,0 +1,149 @@
+package client
+
+import (
+	"fmt"
+	"time"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/state"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+type heartbeatStop struct {
+	lastOk      time.Time
+	allocs      map[string]time.Duration
+	allocHookCh chan *structs.Allocation
+	getRunner   func(string) (AllocRunner, error)
+	logger      hclog.InterceptLogger
+	state       state.StateDB
+}
+
+func newHeartbeatStop(
+	state state.StateDB,
+	getRunner func(string) (AllocRunner, error),
+	logger hclog.InterceptLogger) *heartbeatStop {
+
+	h := &heartbeatStop{
+		allocs:      make(map[string]time.Duration),
+		allocHookCh: make(chan *structs.Allocation),
+		getRunner:   getRunner,
+		logger:      logger,
+		state:       state,
+	}
+
+	if state != nil {
+		lastOk, err := state.GetLastHeartbeatOk()
+		if err == nil && lastOk != nil {
+			h.lastOk = *lastOk
+		}
+	}
+
+	return h
+}
+
+// allocHook is called after (re)storing a new AllocRunner in the client. It registers the
+// allocation to be stopped if the taskgroup is configured appropriately
+func (h *heartbeatStop) allocHook(alloc *structs.Allocation) {
+	tg := allocTaskGroup(alloc)
+	if tg.StopAfterClientDisconnect != nil {
+		h.allocHookCh <- alloc
+	}
+}
+
+// shouldStop is called on a restored alloc to determine if lastOk is sufficiently in the
+// past that it should be prevented from restarting
+func (h *heartbeatStop) shouldStop(alloc *structs.Allocation) bool {
+	tg := allocTaskGroup(alloc)
+	if tg.StopAfterClientDisconnect != nil {
+		now := time.Now()
+		return now.After(h.lastOk.Add(*tg.StopAfterClientDisconnect))
+	}
+	return false
+}
+
+// watch is a loop that checks for allocations that should be stopped. It also manages the
+// registration of allocs to be stopped in a single thread.
```

Comment:
I was going to ask why we didn't push the heartbeatStop down to the individual alloc runners, then I read this. This is a solid "why" comment :+1: